connectivity/check/deployment.go

```diff
@@ -235,6 +243,150 @@ func (ct *ConnectivityTest) deploy(ctx context.Context) error {
 		}
 	}
 
+	// For performance workloads, we want to ensure the client/server are in the same zone
+	// If a zone has > 1 node, use that zone
+	zones := map[string]int{}
+	zone := ""
+	lz := ""
+	n, hasNodes := ct.client.ListNodes(ctx, metav1.ListOptions{})
+	if hasNodes != nil {
+		return fmt.Errorf("unable to query nodes")
+	}
+	for _, l := range n.Items {
+		if _, ok := zones[l.GetLabels()["topology.kubernetes.io/zone"]]; ok {
+			zone = l.GetLabels()["topology.kubernetes.io/zone"]
+			break
+		} else {
+			zones[l.GetLabels()["topology.kubernetes.io/zone"]] = 1
+			lz = l.GetLabels()["topology.kubernetes.io/zone"]
+		}
+	}
+	// No zone had > 1, use the last zone.
+	if zone == "" {
+		ct.Warn("Each zone only has a single node - could impact the performance test results")
+		zone = lz
+	}
+
+	// Need to capture the IP of the Server Deployment, and pass to the client to execute benchmark
+	_, err = ct.clients.src.GetDeployment(ctx, ct.params.TestNamespace, PerfClientDeploymentName, metav1.GetOptions{})
+	if err != nil {
+		ct.Logf("✨ [%s] Deploying Perf Client deployment...", ct.clients.src.ClusterName())
+		perfClientDeployment := newDeployment(deploymentParameters{
+			Name:  PerfClientDeploymentName,
+			Kind:  kindPerfName,
+			Port:  80,
+			Image: defaults.ConnectivityPerformanceImage,
+			Labels: map[string]string{
+				"client": "role",
+			},
+			Command: []string{"/bin/bash", "-c", "sleep 10000000"},
+			Affinity: &corev1.Affinity{
+				NodeAffinity: &corev1.NodeAffinity{
+					PreferredDuringSchedulingIgnoredDuringExecution: []corev1.PreferredSchedulingTerm{
+						{
+							Weight: 100,
+							Preference: corev1.NodeSelectorTerm{
+								MatchExpressions: []corev1.NodeSelectorRequirement{
+									{Key: "topology.kubernetes.io/zone", Operator: corev1.NodeSelectorOpIn, Values: []string{zone}},
+								},
+							},
+						},
+					},
+				},
+			},
+		})
+		_, err = ct.clients.src.CreateDeployment(ctx, ct.params.TestNamespace, perfClientDeployment, metav1.CreateOptions{})
```

Comment:
There's a missing error handling case here, but not sure it makes too much of a difference. The current code is assuming that the way to handle an error from fetching the deployment is to create it, but what if it already exists and it errored during the fetch?

Typically this is handled by doing

```go
if err != nil && k8serrors.IsNotFound(err) {
  // create logic
} else if err != nil {
  // error during fetching
}

https://github.com/kubernetes/apimachinery/blob/2936d3f03931b9c06a641799605d8e806cb2a58b/pkg/api/errors/errors.go#L527

Reply:
Let's discuss this more. I am not sure I am tracking the comment there...

Reply:
From our discussion, https://github.com/cilium/cilium-cli/issues/734